src/typing_manager.ts

```diff
@@ -86,78 +100,60 @@ export class TypingManager implements Disposable {
-                this.typeHandlerDisposable = undefined;
+                this.disposeType();
             }
         } else if (!this.modeManager.isInsertMode) {
             this.isEnteringInsertMode = false;
             this.isExitingInsertMode = false;
-            if (!this.typeHandlerDisposable) {
-                this.logger.debug(`${LOG_PREFIX}: Enabling type handler`);
-                this.typeHandlerDisposable = commands.registerTextEditorCommand("type", this.onVSCodeType);
-            }
+            this.registerType();
         }
     };
 
-    private onVSCodeType = (_editor: TextEditor, edit: TextEditorEdit, type: { text: string }): void => {
-        if (!this.modeManager.isInsertMode || this.modeManager.isRecordingInInsertMode || this.isEnteringInsertMode) {
-            if (this.isEnteringInsertMode) {
-                this.pendingKeysAfterEnter += type.text;
-            } else {
-                this.client.input(normalizeInputString(type.text, !this.modeManager.isRecordingInInsertMode));
-            }
+    private onVSCodeType = async (_editor: TextEditor, edit: TextEditorEdit, type: { text: string }): Promise<void> => {
+        if (this.isEnteringInsertMode) {
+            this.pendingKeysAfterEnter += type.text;
         } else if (this.isExitingInsertMode) {
             this.pendingKeysAfterExit += type.text;
+        } else if (this.modeManager.isInsertMode && !this.modeManager.isRecordingInInsertMode) {
+            const mode = await this.client.mode;
+            if (mode.blocking) {
+                this.client.input(normalizeInputString(type.text, !this.modeManager.isRecordingInInsertMode));
+            } else {
+                this.disposeType();
+                commands.executeCommand("default:type", { text: type.text });
+            }
         } else {
-            commands.executeCommand("default:type", { text: type.text });
+            this.client.input(normalizeInputString(type.text, !this.modeManager.isRecordingInInsertMode));
         }
     };
 
-    private onEscapeKeyCommand = async (): Promise<void> => {
-        this.logger.debug(`${LOG_PREFIX}: Escape key`);
-        if (this.modeManager.isInsertMode) {
-            this.logger.debug(`${LOG_PREFIX}: Syncing buffers with neovim`);
-            this.isExitingInsertMode = true;
-            // rebind early to store fast pressed keys which may happen between sending changes to neovim and exiting insert mode
-            // see https://github.com/asvetliakov/vscode-neovim/issues/324
-            if (!this.typeHandlerDisposable) {
-                this.typeHandlerDisposable = commands.registerTextEditorCommand("type", this.onVSCodeType);
-            }
-            // this.leaveMultipleCursorsForVisualMode = false;
+    private onSyncSendCommand = async (key: string): Promise<void> => {
+        this.logger.debug(`${LOG_PREFIX}: Sync and send for: ${key}`);
+        if (this.modeManager.isInsertMode && !(await this.client.mode).blocking) {
```

Comment:
@justinmk this is the best way I could figure out to use `mode.blocking`. Since it is an async call to nvim, waiting for the result adds delay. I have tried to call await the bare minimum, like in this line where I rely on short-circuit logic. However, this await adds lag in the following situation:
- in insert mode
- Esc/C-o/C-r/C-u/C-w/etc is pressed
- it is important to check whether Esc/C-o will actually exit insert mode or require a resync
- it won't need to exit/resync if it is waiting for a C-r register selection

I don't think the lag is very significant, but basically any insert-mode binding needs to do this check. I think it is ok to merge, however. I was just wondering if there is a better way to obtain `mode.blocking` through UI events or similar, so that the information does not need to be requested on-demand.

Reply:
> basically any insert-mode binding needs to do this check. 

My general opinion remains "who cares about insert-mode bindings?" They should be rare, and vscode-neovim should continue to posit that insert-mode is mostly owned by VSCode, not Nvim, and therefore use of insert-mode bindings are "use at your own peril".

> if there is a better way to obtain mode.blocking through UI events or similar, so that the information does not need to be requested on-demand

 I don't think this makes sense because "is blocking?" is inherently a just-in-time question. Receiving an asynchronous `blocking` flag in a UI event would be pointless and misleading.